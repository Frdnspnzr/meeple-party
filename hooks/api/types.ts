export interface Result<T> {
  isLoading: boolean;
  isError: boolean;
  data: T | undefined | null;
  invalidate: () => void;
}

export interface MutableResult<T> extends Result<T> {
  mutate: (data: Partial<T>, configuration: MutationConfiguration) => void;
}

export interface DeleteableResult<T> extends Result<T> {
  deleteFunction: () => void;
}

export interface MutationConfiguration {}
