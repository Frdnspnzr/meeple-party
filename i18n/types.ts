export type RealLanguage = "de" | "en";
export type PageLanguage = RealLanguage | "auto";
export type GameLanguage = PageLanguage | "follow";
export type Language = RealLanguage | "auto" | "follow";
