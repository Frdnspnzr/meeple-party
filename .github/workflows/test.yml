name: Test
on:
  pull_request:
    branches:
      - "*"
  workflow_dispatch:
    inputs:
      update-snapshots:
        description: "Update snapshots?"
        type: boolean
permissions:
  contents: read
  pages: write
  id-token: write
jobs:
  build:
    timeout-minutes: 10
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: lts/*
          cache: "npm"
      - name: Install dependencies
        run: npm install
      - name: Generate database bindings
        run: npm run "prisma generate"
      - name: Build app
        run: npm run build
      - name: Save build
        uses: actions/upload-artifact@v4
        with:
          name: build
          path: .next

  e2e:
    timeout-minutes: 60
    runs-on: ubuntu-latest
    needs: build
    services:
      postgres:
        image: postgres
        env:
          POSTGRES_PASSWORD: postgres
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 5432:5432
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: lts/*
          cache: "npm"
      - name: Install dependencies
        run: npm install
      - name: Install Playwright browsers
        run: npx playwright install --with-deps
      - name: Load build
        uses: actions/download-artifact@v4
        with:
          name: build
          path: .next
      - name: Set up snapshot cache
        id: snapshots_cache
        uses: actions/cache@v4
        with:
          key: ${{ runner.os }}-snapshots-${{github.ref}}
          restore-keys: ${{ runner.os }}-snapshots-refs/heads/main
          path: ./tests/e2e/snapshots/**
      - name: Update snapshots
        if: ${{steps.snapshots_cache.outputs.cache-hit != 'true' || inputs.update-snapshots == 'true'}}
        run: npx playwright test --update-snapshots
      - name: Generate database bindings
        run: npm run "prisma generate"
      - name: Generate database tables
        run: npm run "prisma deploy"
        env:
          DATABASE_URL: postgresql://postgres:postgres@localhost:5432/test
      - name: Run Playwright tests
        run: npx playwright test --max-failures=10
        env:
          DATABASE_URL: postgresql://postgres:postgres@localhost:5432/test
          NEXT_PUBLIC_SECRET: Testrunner
          NEXTAUTH_URL: http://127.0.0.1:3000
      - uses: actions/upload-artifact@v4
        if: always()
        with:
          name: playwright-report
          path: playwright-report/

  unittest:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: lts/*
          cache: "npm"
      - name: Install dependencies
        run: npm install
      - name: Generate database bindings
        run: npm run "prisma generate"
      - name: Run the tests
        run: npm test
      - name: Upload coverage reports to Codecov
        uses: codecov/codecov-action@v3
        env:
          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}

  publish-e2e:
    runs-on: ubuntu-latest
    needs: [e2e, build]
    if: ${{ !cancelled() && (needs.build.result=='success') }}
    steps:
      - name: Configure GitHub Pages
        uses: actions/configure-pages@v3
      - name: Load report
        uses: actions/download-artifact@v4
        with:
          name: playwright-report
          path: playwright-report/
      - name: Upload Pages Artifact
        uses: actions/upload-pages-artifact@v2
        with:
          path: playwright-report/
      - name: Deploy to GitHub Pages
        uses: actions/deploy-pages@v2
