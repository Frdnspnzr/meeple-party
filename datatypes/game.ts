import * as client from "@prisma/client";

/**
 * @deprecated This type should be replaced by the Data Layer V2 version
 */
export type Game = Omit<client.Game, "updatedAt">;

/**
 * @deprecated This type should be replaced by the Data Layer V2 version
 */
export interface BggGame extends Game {
  description: string;
  designers: string[];
  artists: string[];
  age: number;
}
