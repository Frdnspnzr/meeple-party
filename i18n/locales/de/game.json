{
    "Page": {
        "Header": {
            "Design": "Design",
            "Art": "Illustration",
            "Players": "Spielende",
            "Minutes": "Minuten",
            "Weight": "Komplexität"
        },
        "Collections": {
            "YourCollection": "Deine Sammlung",
            "FriendsOwn": "Im Besitz",
            "FriendsWantToPlay": "Wollen spielen",
            "FriendsWishlist": "Wünschen sich"
        }
    }
}