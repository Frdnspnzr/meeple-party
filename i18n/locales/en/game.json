{
    "Attributes": {
        "Players": "Players",
        "PlayingTime": "Playing Time",
        "Weight": "Weight"
    },
    "Page": {
        "Header": {
            "Design": "Design",
            "Art": "Art",
            "Players": "players",
            "Minutes": "minutes",
            "Weight": "weight"
        },
        "Collections": {
            "YourCollection": "Your collection",
            "FriendsOwn": "Owned by",
            "FriendsWantToPlay": "Want to play",
            "FriendsWishlist": "Wishlist"
        },
        "More": "â€¦ and {{count}} more"
    }
}