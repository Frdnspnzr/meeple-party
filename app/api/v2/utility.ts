import { Relationship, RelationshipType } from "@/datatypes/relationship";
import {
  UserPreferences,
  UserProfile,
  defaultUserPreferences,
} from "@/datatypes/userProfile";
import {
  Prisma,
  Relationship as PrismaRelationship,
  User,
} from "@prisma/client";

export type FullPrismaRelationship = PrismaRelationship & {
  sender: User;
  recipient: User;
};

export function convertToUserProfile(
  user: User,
  friend: boolean = false
): UserProfile {
  return {
    id: user.id,
    name: user.name,
    role: user.role,
    image: user.image,
    realName: allowRealName(user, friend) ? user.realName : null,
    about: user.about,
    place: allowPlace(user, friend) ? user.place : null,
    bggName: allowBggName(user, friend) ? user.bggName : null,
  };
}

export function normalizeRelationship(
  prismaRelationship: FullPrismaRelationship,
  userId: string
): Relationship {
  return {
    profile: getProfile(prismaRelationship, userId),
    type: getRelationshipType(prismaRelationship, userId),
    lastUpdate: prismaRelationship.updatedAt,
  };
}

function allowRealName(user: User, friend: boolean): boolean {
  if (friend) {
    return true;
  }
  const preferences = getPreferences(user);
  return preferences.showRealNameInProfile;
}

function allowPlace(user: User, friend: boolean): boolean {
  if (friend) {
    return true;
  }
  const preferences = getPreferences(user);
  return preferences.showPlaceInProfile;
}

function allowBggName(user: User, friend: boolean): boolean {
  return friend;
}

function getPreferences(user: User): UserPreferences {
  return {
    ...defaultUserPreferences,
    ...(user.preferences as Prisma.JsonObject),
  };
}

function getRelationshipType(
  relationship: FullPrismaRelationship,
  userId: string
): RelationshipType {
  if (relationship.type === "FRIENDSHIP") {
    return RelationshipType.FRIENDSHIP;
  } else if (relationship.type === "FRIEND_REQUEST") {
    if (relationship.sender.id === userId) {
      return RelationshipType.FRIEND_REQUEST_SENT;
    } else {
      return RelationshipType.FRIEND_REQUEST_RECEIVED;
    }
  }
  throw new Error("Error determining relationship type");
}

function getProfile(
  relationship: FullPrismaRelationship,
  userId: string
): UserProfile {
  const profile =
    relationship.recipientId === userId
      ? relationship.sender
      : relationship.recipient;
  return convertToUserProfile(profile, relationship.type === "FRIENDSHIP");
}

/**
 * Reads the authentication token for cronjobs from the environment variable `CRON_AUTH_TOKEN`. Throws when that variable is not set.
 * @returns Authentication token.
 * @throws Throws when no token is configured.
 */
export function getCronAuthToken() {
  if (!process.env.CRON_AUTH_TOKEN) {
    throw new Error("No cron auth token configured for current environment");
  }
  return process.env.CRON_AUTH_TOKEN;
}
