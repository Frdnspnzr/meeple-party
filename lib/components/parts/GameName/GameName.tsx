"use client";

import { ExpandedGame } from "@/datatypes/game";
import usePreferredGameLanguage from "@/lib/hooks/usePreferredGameLanguage";

interface GameNameProps {
  game: ExpandedGame;
}

//TODO Test, document
export default function GameName({ game }: GameNameProps) {
  const { loading, preferredGameLanguage } = usePreferredGameLanguage();

  const translatedName = game.names.find(
    (n) => n.language === preferredGameLanguage
  );

  if (loading || !translatedName) {
    return <>{game.name}</>;
  } else {
    return <>{translatedName.name}</>;
  }
}
