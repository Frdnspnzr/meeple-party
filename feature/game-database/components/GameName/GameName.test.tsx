import usePreferredGameLanguage from "@/lib/hooks/usePreferredGameLanguage";
import { gameToExpandedGame } from "@/lib/types/game";
import { generateGame } from "@/lib/utility/test";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import GameName from "./GameName";

jest.mock("@/lib/hooks/usePreferredGameLanguage");

describe("Part: GameName", () => {
  it("matches the snapshot for game without additional names", async () => {
    jest
      .mocked(usePreferredGameLanguage)
      .mockReturnValueOnce({ loading: false, preferredGameLanguage: "en" });
    const game = gameToExpandedGame(generateGame(1));
    game.name = "Static game name";
    const { container } = render(<GameName game={game} />);

    expect(container).toMatchSnapshot();
  });
  it("matches the snapshot for game without fitting alternate name", async () => {
    jest
      .mocked(usePreferredGameLanguage)
      .mockReturnValueOnce({ loading: false, preferredGameLanguage: "en" });
    const game = gameToExpandedGame(generateGame(1));
    game.name = "Static game name";
    game.names.push({ language: "de", name: "Alternate static game name" });
    const { container } = render(<GameName game={game} />);

    expect(container).toMatchSnapshot();
  });
  it("matches the snapshot for game with alternate name", async () => {
    jest
      .mocked(usePreferredGameLanguage)
      .mockReturnValueOnce({ loading: false, preferredGameLanguage: "de" });
    const game = gameToExpandedGame(generateGame(1));
    game.name = "Static game name";
    game.names.push({ language: "de", name: "Alternate static game name" });
    const { container } = render(<GameName game={game} />);

    expect(container).toMatchSnapshot();
  });
});
