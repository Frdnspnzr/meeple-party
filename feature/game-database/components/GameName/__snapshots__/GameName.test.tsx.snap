// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Part: GameName matches the snapshot for game with alternate name 1`] = `
<div>
  Alternate static game name
</div>
`;

exports[`Part: GameName matches the snapshot for game without additional names 1`] = `
<div>
  Static game name
</div>
`;

exports[`Part: GameName matches the snapshot for game without fitting alternate name 1`] = `
<div>
  Static game name
</div>
`;
